doc/a2m.1
.TH A2M 1 "26 October 2001" "WFDB software 10.2" "WFDB applications"
.SH NAME
a2m, ad2m, m2a, md2a \- utilities for converting between MIT and AHA DB formats
.SH SYNOPSIS
\fBa2m -i \fIahafile\fB -r \fIrecord\fB -a \fIannotator\fR [ \fIoptions\fR ... ]
.br
\fBad2m -i \fIahafile\fB -r \fIrecord\fR [ \fIoptions\fR ... ]
.br
\fBm2a -r \fIrecord\fB -a \fIMIT-annotator AHA-annotator\fR [ \fIoptions\fR ... ]
.br
\fBmd2a -o \fIahafile\fB -r \fIrecord\fR [ \fIoptions\fR ... ]
.SH DESCRIPTION
These programs can be used to convert AHA DB distribution files from half-inch
9-track tape or floppy disks to MIT format (i.e., formats that can be read
directly by programs compiled with the \fIwfdb\fR(3) library), or to convert
files in MIT format into AHA DB tape distribution format (conversion to
AHA DB floppy disk distribution format is not supported).  All four of these
programs print a brief usage summary if invoked with no command-line arguments,
or with a \fB-h\fR option.
.PP
Use \fIa2m\fR to convert AHA-format annotation files into MIT format.  Options
for \fIa2m\fR include:
.TP
\fB-s\fI time\fR
Shift annotations forward by the specified \fItime\fR (default: no shift for
type 0 input files, 5 minutes for type 1 or 3, or 2 hours and 30 minutes for
type 2).
.TP
\fB-t\fI type\fR
Convert an input file of the specified \fItype\fR (0:
a file produced by a DB application using \fIputann\fR and
`WFDB_AHA_WRITE' mode; 1: an AHA DB `short format' tape file; 2: an
AHA DB `long format' tape file; 3: an AHA DB compressed (\fI*.ano\fR)
floppy disk file).  Input files of types 1, 2, and 3 are assumed to
contain annotation times in milliseconds, which are converted to
sampling intervals based on an assumed sampling frequency of 250 Hz.
.PP
Use \fIad2m\fR to convert AHA-format signal files into MIT format.  Options
for \fIad2m\fR include:
.TP
\fB-c\fR
Convert an AHA DB compressed (\fI*.cmp\fR) floppy disk file (default: convert a
file in AHA DB tape format).
.TP
\fB-f\fI time\fR
Begin converting at the specified \fItime\fR relative to the beginning of the
input file (default: 0, i.e., at the beginning of the input file)
.TP
\fB-t\fI time\fR
Stop converting at the specified \fItime\fR relative to the beginning of the
input file (default: 35 minutes after the starting time, or the end of the
input file, whichever comes first).
.PP
Use \fIm2a\fR to convert MIT-format annotation files into AHA tape format.
Options for \fIm2a\fR include:
.TP
\fB-s\fI time\fR
Shift annotation times backward by the specified \fItime\fR, and convert them
from sample intervals to milliseconds.
.PP
Use \fImd2a\fR to convert MIT-format signal files into AHA tape format.
Options for \fImd2a\fR include:
.TP
\fB-n\fI new-record\fR
Create a new header file for the AHA-format output signal file, so that it
may be read as record \fInew-record\fR.
.PP
MIT-format files can be excerpted and reformatted in more generally useful ways
using \fIsnip\fR(1) or \fIxform\fR(1).
.PP
The shell variable \fBWFDB\fR should be set and exported (see
\fIsetwfdb\fR(1)).
.SH EXAMPLES
.SS "AHA DB floppy disk"
To make a version of AHA DB record 1201 in MIT format, given the distribution
floppy disk, copy the files \fI1201.ano\fR and \fI1201.cmp\fR to the current
directory, then type:
.br
	ad2m -i 1201.cmp -r 1201 -c
.br
	a2m -i 1201.ano -r 1201 -a atr -t 3
.br
These commands produce files \fI1201.dat\fR (the signal file), \fI1201.hea\fR
(the header file), and \fI1201.atr\fR (the reference annotation file), all in
the current directory.  Run \fIad2m\fR first, so that the new header file is
available for the use of \fIa2m\fR.
.SS "AHA DB short format tape"
.PP
To obtain the same files given a `short format' 9-track distribution tape,
copy the second and third files from the tape into files \fI1201.tap\fR and
\fI1201.ann\fR in the current directory, then type:
.br
	ad2m -i 1201.tap -r 1201
.br
	a2m -i 1201.ann -r 1201 -a atr -t 1
.br
The names for the files copied from the tape are arbitrary, but do not
use names of files to be generated by \fIad2m\fR or \fIa2m\fR (see the
previous example).  Note that the first and fourth files on the
distribution tape contain an `id' block, which can be read by
\fIreadid\fR (a program included in the \fIconvert\fR directory of the
WFDB Software Package) to verify the record name. Distribution tapes
that contain more than one record contain additional sets of four
files, always in the same order within each set.
.SS "AHA DB long format tape"
.PP
To make a version of the three-hour AHA DB record 1001 in MIT format,
given the `long format' distribution tape, copy the second and third files
from the tape into files \fI1001.tap\fR and \fI1001.ann\fR in the current
directory, then type:
.br
	ad2m -i 1001.tap -r 1001 -t 3:0:0
.br
	a2m -i 1001.ann -r 1001 -a atr -t 2
.br
The \fB-t 3:0:0\fR option is necessary to prevent \fIad2m\fR from truncating
the signal file after the first 35 minutes.
.SS "Converting AHA DB long format to short format"
.PP
To make a version of AHA DB record 1201 in MIT format, given a `long format'
9-track distribution tape containing the corresponding three-hour record 1001,
copy the second and third files from the tape into files \fI1001.tap\fR and
\fI1001.ann\fR in the current directory, then type:
.br
	ad2m -i 1001.tap -r 1201 -f 2:25:0
.br
	a2m -i 1001.ann -r 1201 -a atr -t 1
.br
In this case, the \fB-f\fR option instructs \fIad2m\fR to skip the first
two hours and 25 minutes of the `long-format' AHA signal file, and to reformat
the remainder (equivalent to the 35-minute `short-format' record).  The
\fB-t 1\fR option is used with \fIa2m\fR even though its input file comes from
a `long-format' tape, because the annotation times must be shifted only by
the amount necessary for a `short-format' tape in this case.
.SS "Sharing signal files for long format and short format AHA DB records"
.PP
To keep both versions (1001 and 1201) on-line, make the `long format' version
first (see above), then type:
.br
	a2m -i 1001.ann -r 1201 -a atr -t 1
.br
to make a `short format' reference annotation file.  Continue (under UNIX) by:
.br
	cp 1001.hea 1201.hea
.br
or (under MS-DOS) by:
.br
	copy 1001.hea 1201.hea
.br
and edit the copy of the header file, replacing `1001' in the first line
(only!) with `1201', and replacing `212' in the second and third lines by
`212+6525000' (see the description of the `byte offset' field in
\fIheader\fR(5)).  Although two header and reference annotation files are
needed, both versions can share the same signal file, allowing a substantial
savings in storage requirements.  Note that WFDB application
programs that read the `short format' record 1201 signal file may report
signal checksum errors at the end of the record, unless you also recalculate
the signal checksums (easily done using \fIsnip\fR(1) to copy the record;
delete the copy once the checksums have been obtained).
.SH AVAILABILITY
These programs are provided in the \fIconvert\fR directory of the WFDB Software
Package.  Run \fImake\fR in that directory to compile and install them if they
have not been installed already.
.SH SEE ALSO
snip(1), xform(1), wfdb(3), header(5)
.SH AUTHOR
George B. Moody (george@mit.edu)
.SH SOURCES
http://www.physionet.org/physiotools/wfdb/convert/a2m.c
.br
http://www.physionet.org/physiotools/wfdb/convert/ad2m.c
.br
http://www.physionet.org/physiotools/wfdb/convert/m2a.c
.br
http://www.physionet.org/physiotools/wfdb/convert/md2a.c


src/convert/a2m.c
/* file: a2m.c		G. Moody        9 June 1983
			Last revised:  27 July 2010

-------------------------------------------------------------------------------
a2m: Convert an AHA format annotation file to MIT format
Copyright (C) 1983-2010 George B. Moody

This program is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation; either version 2 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program; if not, see <http://www.gnu.org/licenses/>.

You may contact the author by e-mail (wfdb@physionet.org) or postal mail
(MIT Room E25-505A, Cambridge, MA 02139 USA).  For updates to this software,
please visit PhysioNet (http://www.physionet.org/).
_______________________________________________________________________________

This program converts four types of AHA format annotation files to MIT format:

 Type	Description

   0	AHA-format annotation files produced by WFDB library-based
	programs that use AHA_WRITE mode and `putann' (the `m2a'
	translator is an example of such a program).  Time is
	measured in sample intervals from the beginning of the record,
	and MIT annotation codes are included and read into the `chan'
	field of the annotation by `getann';  if such codes are
	present (i.e., if `chan' is non-zero), the AHA code is
	ignored.

   1	Standard AHA annotation files from short-format AHA DB
	distribution tapes.  Time is measured in milliseconds from
	the beginning of the test period (five minutes, or 75000 sample
	intervals at 4 ms/sample interval, after the first sample).

   2	Standard AHA annotation files from long-format AHA DB
	distribution tapes.  Time is measured as for filetype 1,
	except that the test period begins 2.5 hours (2,225,000 sample
	intervals) after the first sample.

   3	`Compressed' (*.ANO) annotation files from AHA DB CD-ROMs or floppy
        diskettes.  Time is measured as for files of type 1 or 2, depending
	on the record name (x2xx or x3xx: short format, x0xx or x1xx: long
	format).
*/

#include <stdio.h>
#ifdef __STDC__
#define RB	"rb"
#else
#ifdef MSDOS
#define RB	"rb"
#else
#define RB	"r"
extern void exit();
#endif
#endif

#include <wfdb/wfdb.h>
#define isqrs
#define map1
#define map2
#define mamap
#define annpos
#include <wfdb/ecgmap.h>

char *pname;

main(argc, argv)
int argc;
char *argv[];
{
    char *ifname = NULL, *oann = "atr", *p, *record = NULL, *prog_name();
    int i, type = -1, getcann();
    long offset = 0L;
    unsigned int nann = 2;
    WFDB_Anninfo afarray[2];
    static WFDB_Annotation annot;
    void help();

    /* Read and interpret command-line arguments. */
    pname = prog_name(argv[0]);
    for (i = 1; i < argc; i++) {
	if (*argv[i] == '-') switch (*(argv[i]+1)) {
	  case 'a':	/* annotator name follows */
	    if (++i >= argc) {
		(void)fprintf(stderr,
			      "%s: output annotator name must follow -a\n",
			      pname);
		exit(1);
	    }
	    oann = argv[i];
	    break;
	  case 'h':	/* help requested */
	    help();
	    exit(1);
	    break;
	  case 'i':	/* input file name follows */
	    if (++i >= argc) {
		(void)fprintf(stderr,
		      "%s: name of AHA-format input file must follow -i\n",
			      pname);
		exit(1);
	    }
	    ifname = argv[i];
	    p = ifname + strlen(ifname) - 4;
	    if (type < 0 && p > ifname && 
		(strcmp(p, ".ANO") == 0 || strcmp(p, ".ano") == 0))
		type = 3;
	    break;
	  case 'r':	/* record name follows */
	    if (++i >= argc) {
		(void)fprintf(stderr, "%s: record name must follow -a\n",
			      pname);
		exit(1);
	    }
	    record = argv[i];
	    break;
	  case 's':	/* time shift follows */
	    if (++i >= argc) {
		(void)fprintf(stderr, "%s: time shift must follow -s\n",
			      pname);
		exit(1);
	    }
	    offset = i;	/* save position in arg list, convert to time
			   later when sampling frequency is known */
	    break;
	  case 't':	/* file type follows */
	    if (++i >= argc || (type = atoi(argv[i])) < 0 || type > 3) {
		(void)fprintf(stderr, "%s: file type (0-3) must follow -t\n",
			      pname);
		exit(1);
	    }
	    break;
	  default:
	    (void)fprintf(stderr, "%s: unrecognized option %s\n", pname,
			  argv[i]);
	    exit(1);
	}
	else {
	    (void)fprintf(stderr, "%s: unrecognized argument %s\n", pname,
			  argv[i]);
	    exit(1);
	}
    }

    /* Quit if input filename was not specified. */
    if (ifname == NULL) {
	help();
	exit(1);
    }
		 
    /* Attach the input file to the standard input. */
    if (strcmp(ifname, "-") != 0 && freopen(ifname, RB, stdin) == NULL) {
	(void)fprintf(stderr, "%s: can't read input file `%s'\n", pname,
		      ifname);
	exit(2);
    }

    /* If no output record name was specified, use the basename of the input
       file (prog_name strips off any path info from ifname, and any extension
       is stripped off in the 'for' loop. */
    if (record == NULL) {
	record = prog_name(ifname);
	for (p = record+1; *p; p++)
	    if (*p == '.') {
		*p = '\0';
		break;
	    }
    }

    if (type < 0) type = 0;

    /* Open the annotation files. */
    afarray[0].name = oann;
    afarray[0].stat = WFDB_WRITE;
    if (type != 3) {
	afarray[1].name = "-";
	afarray[1].stat = WFDB_AHA_READ;
    }
    else
	nann = 1;
    if (annopen(record, afarray, nann) < 0)
	exit(2);

    /* Set the time shift. */
    if (sampfreq(record) == 0.)
	(void)setsampfreq(250.);
    if (offset) offset = strtim(argv[offset]);
    else {
	switch (type) {
	  case 0:
	    break;
	  case 1:
	    offset = strtim("5:0");
	    break;
	  case 2:
	    offset = strtim("2:30:0");
	    break;
	  case 3:
	    if (strlen(record) == 4 && (record[1] == '0' || record[1] == '1'))
		offset = strtim("2:30:0");  /* long format */
	    else if (strlen(record) == 4 && (record[1] =='2'||record[1]=='3'))
		offset = strtim("5:0");	/* short format */
	    /* otherwise, leave offset at zero (unrecognized format) */
	    break;
	}
    }

    /* Reformat the annotations. */
    switch (type) {
      case 0:
	while (getann(0, &annot) >= 0) {
	    if (annot.chan) {	/* MIT annotation code present */
		annot.anntyp = annot.chan;
		annot.chan = 0;
	    }
	    annot.time += offset;
	    if (putann(0, &annot) < 0)
		break;
	}
	break;
      case 1:
      case 2:
	while (getann(0, &annot) >= 0) {
	    annot.time = (annot.time >> 2) + offset;	/* 4 msec/sample */
	    (void)putann(0, &annot);
	}
	break;
      case 3:
	while (getcann(&annot) >= 0) {
	    annot.time = (annot.time >> 2) + offset;	/* 4 msec/sample */
	    (void)putann(0, &annot);
	}
	break;
    }

    wfdbquit();
    exit(0);	/*NOTREACHED*/
}

int getcann(ap)
WFDB_Annotation *ap;
{
    char buf[6];
    long h, m, l;

    if (fread(buf, 1, 6, stdin) != 6) {		/* hard EOF */
/*	(void)fprintf(stderr, "unexpected EOF in input file\n");	*/
/*	The warning message was removed because the CD-ROM version of these
	files does not include the soft EOF as on the floppies. */
	return (-1);
    }
    /* Compressed AHA format:
       buf[0]	AHA label (`N', `V', `F', `R', `E', `P', `Q', `[', `]', or `U')
       buf[1]	high byte of ahatime, or 0xff if at soft EOF
       buf[2]	middle byte of ahatime
       buf[3]	low byte of ahatime
       buf[4]	high byte of annotation serial number (ignored here)
       buf[5]	low byte of serial number
     */
    
    if ((unsigned)(buf[1] & 0xff) == 0xff)	/* soft EOF */
	return (-1);

    ap->anntyp = ammap(buf[0]);	     /* convert AHA label to annotation code */

    if (ap->anntyp == 0)
	return (-1);		     /* alternate soft EOF, used in a few AHA
					DB CD-ROM annotation files */
    if (buf[0] == 'U')		     /* AHA `U' (unreadable data) label */
	ap->subtyp = -1;		/* decode as NOISE, subtype = -1 */
    else
	ap->subtyp = 0;			/* all others have subtype = 0 */

    h = (long)buf[1] & 0xff;
    m = (long)buf[2] & 0xff;
    l = (long)buf[3] & 0xff;
    ap->time = (h << 16) | (m << 8) | l; /* time in milliseconds from the
					   start of the annotated segment */
    return (0);
}

char *prog_name(s)
char *s;
{
    char *p = s + strlen(s);

#ifdef MSDOS
    while (p >= s && *p != '\\' && *p != ':') {
	if (*p == '.')
	    *p = '\0';		/* strip off extension */
	if ('A' <= *p && *p <= 'Z')
	    *p += 'a' - 'A';	/* convert to lower case */
	p--;
    }
#else
    while (p >= s && *p != '/')
	p--;
#endif
    return (p+1);
}

static char *help_strings[] = {
  "usage: %s -i AHAFILE [ OPTIONS ... ]\n",
  "where AHAFILE is the name of the AHA-format input annotation file,",
  "and OPTIONS may include:",
  " -a ANNOTATOR  specify the output annotator name (default: 'atr')",
  " -h            print this usage summary",
  " -r RECORD     specify the name of the output RECORD (default: use the",
  "                input file name, less path info and extension, as RECORD)",
  " -s TIME       shift annotations by TIME",
  "                (defaults: 0 for TYPE 0,",
  "                      5:0 for TYPE 1, (or TYPE 3 if RECORD=x2xx or x3xx)",
  "                   2:30:0 for TYPE 2, (or TYPE 3 if RECORD=x0xx or x1xx)",
  " -t TYPE       specify type of file to be converted:",
  "                0 for files produced by WFDB applications in AHA format,",
  "                1 for short-format tape files,",
  "                2 for long-format tape files, or",
  "                3 for compressed files from AHA DB CD-ROMs or diskettes",
  "                (defaults: 3 if AHAFILE ends in .ANO or .ano, else 0)",
  NULL
};

void help()
{
    int i;

    (void)fprintf(stderr, help_strings[0], pname);
    for (i = 1; help_strings[i] != NULL; i++)
	(void)fprintf(stderr, "%s\n", help_strings[i]);
}
